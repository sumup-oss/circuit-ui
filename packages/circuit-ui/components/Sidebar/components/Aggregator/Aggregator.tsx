/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { useState, useEffect, ReactElement } from 'react';
import { css } from '@emotion/react';
import { TrackingElement } from '@sumup/collector';

import { ClickEvent } from '../../../../types/events';
import { isEmpty } from '../../../../util/helpers';
import styled, { StyleProps } from '../../../../styles/styled';
import { useClickEvent, TrackingProps } from '../../../../hooks/useClickEvent';
import { Child, hasSelectedChild, getIcon } from '../../SidebarService';
import { SubNavList } from '../SubNavList';
import { NavLabel } from '../NavLabel';
import { AccessibilityError } from '../../../../util/errors';

export interface AggregatorProps {
  /**
   * The children component passed to the NavAggregator
   */
  children: Child | Child[];
  /**
   * The label of NavAggregator
   */
  label: string;
  /**
   * The icon to be shown when the NavAggregator is not selected
   */
  defaultIcon?: ReactElement;
  /**
   * The icon to be shown when the NavAggregator is selected
   */
  selectedIcon?: ReactElement;
  /**
   * Disables the Aggregator and all children
   */
  disabled?: boolean;
  /**
   * The onClick method to handle the click event on the NavAggregator
   */
  onClick?: (event: ClickEvent) => void;
  /**
   * @deprecated
   *
   * Use an `onClick` handler to dispatch user interaction events instead.
   */
  tracking?: TrackingProps;
}

type Disabled = { disabled?: boolean };
type Selected = { selected: boolean };

const baseStyles = ({ theme }: StyleProps) => css`
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  height: auto;
  width: calc(100% - (2 * ${theme.spacings.mega}));
  margin: ${theme.spacings.mega};
  padding: ${theme.spacings.bit};
  cursor: pointer;
  color: ${theme.colors.n300};
  transition: color ${theme.transitions.default};
  font: inherit;
  background: none;
  border: 0;
  outline: none;
`;

const hoverStyles = ({
  theme,
  disabled,
  selected,
}: StyleProps & Disabled & Selected) =>
  !disabled &&
  !selected &&
  css`
    &:hover,
    &:focus {
      color: ${theme.colors.n100};
    }
  `;

const selectedStyles = ({ theme, selected }: StyleProps & Selected) =>
  selected &&
  css`
    color: ${theme.colors.white};
  `;

const disabledStyles = ({ theme, disabled }: StyleProps & Disabled) =>
  disabled &&
  css`
    cursor: not-allowed;
    color: ${theme.colors.n500};
  `;

const AggregatorContainer = styled('button')<Disabled & Selected>(
  baseStyles,
  hoverStyles,
  selectedStyles,
  disabledStyles,
);

const TRACKING_ELEMENT = 'aggregator';

export function Aggregator({
  children,
  label,
  defaultIcon,
  selectedIcon,
  disabled,
  onClick,
  tracking,
  ...props
}: AggregatorProps): JSX.Element {
  if (
    process.env.NODE_ENV !== 'production' &&
    process.env.NODE_ENV !== 'test' &&
    !label
  ) {
    throw new AccessibilityError('Aggregator', 'The `label` prop is missing.');
  }
  const [isOpen, setIsOpen] = useState(false);
  const selectedChild = hasSelectedChild(children);
  const baseHandleClick = (event: ClickEvent) => {
    if (onClick) {
      onClick(event);
    }

    if (selectedChild) {
      return;
    }

    setIsOpen((open) => !open);
  };
  const handleClick = useClickEvent(
    baseHandleClick,
    tracking,
    'sidebar-nav-aggregator',
  );

  useEffect(() => {
    setIsOpen(selectedChild);
  }, [selectedChild]);

  const icon = getIcon({
    selected: selectedChild,
    selectedIcon,
    defaultIcon,
    disabled,
  });

  return (
    <li>
      <AggregatorContainer
        selected={selectedChild}
        disabled={disabled}
        aria-disabled={disabled}
        onClick={handleClick}
        {...props}
      >
        {icon}
        <NavLabel secondary={false}>{label}</NavLabel>
      </AggregatorContainer>
      {!isEmpty(children) && !disabled && (
        <TrackingElement name={TRACKING_ELEMENT}>
          <SubNavList visible={isOpen}>{children}</SubNavList>
        </TrackingElement>
      )}
    </li>
  );
}
