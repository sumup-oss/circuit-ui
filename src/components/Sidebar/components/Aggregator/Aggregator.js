/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import { css } from '@emotion/core';

import SubNavList from '../SubNavList';
import NavLabel from '../NavLabel';
import { hasSelectedChild, getIcon } from '../NavItem/utils';
import { childrenPropType } from '../../../../util/shared-prop-types';

const baseStyles = ({ theme }) => css`
  label: nav-aggregator;
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  height: auto;
  margin: ${theme.spacings.mega};
  padding: ${theme.spacings.bit};
  cursor: pointer;
  color: ${theme.colors.n500};
`;

const hoverStyles = ({ theme, disabled, selected }) =>
  !disabled &&
  !selected &&
  css`
    label: nav-aggregator--hover;
    &:hover {
      color: ${theme.colors.n300};
    }
  `;

const selectedStyles = ({ theme, selected }) =>
  selected &&
  css`
    label: nav-aggregator--active;
    color: ${theme.colors.n100};
  `;

const disabledStyles = ({ theme, disabled }) =>
  disabled &&
  css`
    label: nav-item--disabled;
    cursor: not-allowed;
    color: ${theme.colors.n700};
  `;

const AggregatorContainer = styled.div(
  baseStyles,
  hoverStyles,
  selectedStyles,
  disabledStyles
);

class Aggregator extends Component {
  state = { open: false };

  static getDerivedStateFromProps(props) {
    if (hasSelectedChild(props.children)) {
      return { open: true };
    }
    return null;
  }

  componentDidUpdate(prevProps, prevState) {
    const { children } = this.props;
    const shouldClose =
      (hasSelectedChild(prevProps.children) || prevState.open) &&
      !hasSelectedChild(children);

    if (shouldClose) {
      this.closeAggregator();
    }
  }

  closeAggregator = () => {
    this.setState({ open: false });
  };

  toggleAggregator = () => {
    const { onClick } = this.props;

    this.setState(prevState => ({ open: !prevState.open }));
    if (onClick) {
      onClick();
    }
  };

  render() {
    const {
      children,
      label,
      defaultIcon,
      selectedIcon,
      disabled,
      onClick,
      ...props
    } = this.props;
    const { open } = this.state;
    const icon = getIcon({
      selected: hasSelectedChild(children),
      selectedIcon,
      defaultIcon,
      disabled
    });

    return (
      <Fragment>
        <AggregatorContainer
          selected={hasSelectedChild(children)}
          disabled={disabled}
          onClick={this.toggleAggregator}
          {...props}
        >
          {icon}
          <NavLabel>{label}</NavLabel>
        </AggregatorContainer>
        {children && !disabled && (
          <SubNavList visible={open}>{children}</SubNavList>
        )}
      </Fragment>
    );
  }
}

Aggregator.propTypes = {
  /**
   * The children component passed to the NavAggregator
   */
  children: childrenPropType,
  /**
   * The label of NavAggregator
   */
  label: PropTypes.string,
  /**
   * The icon to be shown when the NavAggregator is not selected
   */
  defaultIcon: PropTypes.node,
  /**
   * The icon to be shown when the NavAggregator is selected
   */
  selectedIcon: PropTypes.node,
  /**
   * Disables the Aggregator and all children
   */
  disabled: PropTypes.bool,
  /**
   * The onClick method to handle the click event on the NavAggregator
   */
  onClick: PropTypes.func
};

Aggregator.defaultProps = {
  children: null,
  label: '',
  defaultIcon: '',
  selectedIcon: '',
  disabled: false,
  onClick: null
};

/**
 * @component
 */
export default Aggregator;
