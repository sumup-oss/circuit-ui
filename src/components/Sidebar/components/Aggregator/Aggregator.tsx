/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/** @jsx jsx */
import {
  useState,
  useEffect,
  FC,
  SVGProps,
  ReactNode,
  MouseEvent,
  KeyboardEvent,
} from 'react';
import { css, jsx } from '@emotion/core';
import { TrackingElement } from '@sumup/collector';

import styled, { StyleProps } from '../../../../styles/styled';
import SubNavList from '../SubNavList';
import BaseNavLabel from '../NavLabel';
import { hasSelectedChild, getIcon } from '../NavItem/utils';
import useClickHandler from '../../../../hooks/use-click-handler';

export interface AggregatorProps {
  /**
   * The children component passed to the NavAggregator
   */
  children: ReactNode;
  /**
   * The label of NavAggregator
   */
  label: string;
  /**
   * The icon to be shown when the NavAggregator is not selected
   */
  defaultIcon: FC<SVGProps<SVGSVGElement>>;
  /**
   * The icon to be shown when the NavAggregator is selected
   */
  selectedIcon: FC<SVGProps<SVGSVGElement>>;
  /**
   * Disables the Aggregator and all children
   */
  disabled: boolean;
  /**
   * The onClick method to handle the click event on the NavAggregator
   */
  onClick?: (event: MouseEvent | KeyboardEvent) => void;
  /**
   * Additional data that is dispatched with click tracking event.
   */
  tracking: {
    label?: string;
    component?: string;
    customParameters?: {
      [key: string]: unknown;
    };
  };
}

type Disabled = { disabled: boolean };
type Selected = { selected: boolean };

const baseStyles = ({ theme }: StyleProps) => css`
  label: nav-aggregator;
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  height: auto;
  width: calc(100% - (2 * ${theme.spacings.mega}));
  margin: ${theme.spacings.mega};
  padding: ${theme.spacings.bit};
  cursor: pointer;
  color: ${theme.colors.n300};
  transition: color ${theme.transitions.default};
  font: inherit;
  background: none;
  border: 0;
  outline: none;
`;

const hoverStyles = ({
  theme,
  disabled,
  selected,
}: StyleProps & Disabled & Selected) =>
  !disabled &&
  !selected &&
  css`
    label: nav-aggregator--hover;
    &:hover,
    &:focus {
      color: ${theme.colors.n100};
    }
  `;

const selectedStyles = ({ theme, selected }: StyleProps & Selected) =>
  selected &&
  css`
    label: nav-aggregator--active;
    color: ${theme.colors.white};
  `;

const disabledStyles = ({ theme, disabled }: StyleProps & Disabled) =>
  disabled &&
  css`
    label: nav-item--disabled;
    cursor: not-allowed;
    color: ${theme.colors.n500};
  `;

const NavLabel = BaseNavLabel as any;

const AggregatorContainer = styled('button')<Disabled & Selected>(
  baseStyles,
  hoverStyles,
  selectedStyles,
  disabledStyles,
);

const TRACKING_ELEMENT = 'aggregator';

const Aggregator = ({
  children,
  label,
  defaultIcon,
  selectedIcon,
  disabled,
  onClick,
  tracking,
  ...props
}: AggregatorProps) => {
  const [isOpen, setIsOpen] = useState(false);
  const selectedChild = hasSelectedChild(children);
  const baseHandleClick = (event: MouseEvent | KeyboardEvent) => {
    if (onClick) {
      onClick(event);
    }

    if (selectedChild) {
      return;
    }

    setIsOpen((open) => !open);
  };
  const handleClick = useClickHandler(
    baseHandleClick,
    tracking,
    'sidebar-nav-aggregator',
  );

  useEffect(() => {
    setIsOpen(selectedChild);
  }, [selectedChild]);

  const icon = getIcon({
    selected: selectedChild,
    selectedIcon,
    defaultIcon,
    disabled,
  });

  return (
    <li>
      <AggregatorContainer
        selected={selectedChild}
        disabled={disabled}
        aria-disabled={disabled}
        onClick={handleClick}
        {...props}
      >
        {icon}
        <NavLabel secondary={false}>{label}</NavLabel>
      </AggregatorContainer>
      {children && !disabled && (
        <TrackingElement name={TRACKING_ELEMENT}>
          <SubNavList visible={isOpen}>{children}</SubNavList>
        </TrackingElement>
      )}
    </li>
  );
};

/**
 * @component
 */
export default Aggregator;
