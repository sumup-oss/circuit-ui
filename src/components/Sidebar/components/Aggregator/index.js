import Aggregator from './Aggregator';

export default Aggregator;
