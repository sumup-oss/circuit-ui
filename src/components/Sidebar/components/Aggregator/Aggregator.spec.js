/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { Component, Fragment } from 'react';

import Aggregator from './Aggregator';

const props = {
  'label': 'Aggregator',
  'onClick': jest.fn(),
  'selectedIcon': 'selected-icon',
  'defaultIcon': 'default-icon',
  'data-testid': 'aggregator',
};

class MockedNavigation extends Component {
  state = {
    selected: 1,
  };

  render() {
    const { selected } = this.state;
    return (
      <Fragment>
        <Aggregator {...props} {...this.props}>
          <div selected={selected === 0} data-testid="child">
            child
          </div>
        </Aggregator>
        {/* eslint-disable-next-line */}
        <div
          id="sibling"
          selected={selected !== 0}
          onClick={() => this.setState({ selected: 1 })}
        >
          sibling
        </div>
      </Fragment>
    );
  }
}

describe('Aggregator', () => {
  describe('styles', () => {
    it('should render with default styles', () => {
      const actual = create(
        <Aggregator {...props}>
          <div data-testid="child">child</div>
        </Aggregator>,
      );
      expect(actual).toMatchSnapshot();
    });

    it('should render and match snapshot when open', () => {
      const { container, getByTestId } = render(
        <Aggregator {...props}>
          <div data-testid="child">child</div>
        </Aggregator>,
      );
      const aggregatorEl = getByTestId('aggregator');

      act(() => {
        fireEvent.click(aggregatorEl);
      });

      expect(container.children).toMatchSnapshot();
    });

    it('should render with disabled state styles and match the snapshot', () => {
      const actual = create(
        <Aggregator {...props} disabled={true}>
          <div data-testid="child">child</div>
        </Aggregator>,
      );

      expect(actual).toMatchSnapshot();
    });
  });

  describe('interactions', () => {
    it('should show children and call onClick when clicking the aggregator', () => {
      const onClick = jest.fn();
      const { getByTestId } = render(
        <Aggregator {...props} onClick={onClick}>
          <div data-testid="child">child</div>
        </Aggregator>,
      );
      const aggregatorEl = getByTestId('aggregator');
      const childEl = getByTestId('child');

      expect(childEl).not.toBeVisible();

      act(() => {
        fireEvent.click(aggregatorEl);
      });

      expect(childEl).toBeVisible();
      expect(onClick).toHaveBeenCalledTimes(1);
    });

    it('should show children when clicking the aggregator and no onClick handle is passed', () => {
      const noHandlerProps = {
        ...props,
        onClick: null,
      };

      const { getByTestId } = render(
        <Aggregator {...noHandlerProps}>
          <div data-testid="child">child</div>
        </Aggregator>,
      );
      const aggregatorEl = getByTestId('aggregator');
      const childEl = getByTestId('child');

      expect(childEl).not.toBeVisible();

      act(() => {
        fireEvent.click(aggregatorEl);
      });

      expect(childEl).toBeVisible();
    });

    it('should not toggle when clicking again on the aggregator with a selected child', () => {
      const onClick = jest.fn();
      const { getByTestId } = render(
        <Aggregator {...props} onClick={onClick}>
          <div selected data-testid="child">
            child
          </div>
        </Aggregator>,
      );
      const aggregatorEl = getByTestId('aggregator');
      const childEl = getByTestId('child');

      expect(childEl).toBeVisible();

      act(() => {
        fireEvent.click(aggregatorEl);
      });

      expect(childEl).toBeVisible();

      act(() => {
        fireEvent.click(aggregatorEl);
      });

      expect(onClick).toHaveBeenCalledTimes(2);
      expect(childEl).toBeVisible();
    });

    it('should close when there are no selected children', () => {
      const onClick = jest.fn();
      const { getByTestId } = render(<MockedNavigation onClick={onClick} />);
      const aggregatorEl = getByTestId('aggregator');
      const childEl = getByTestId('child');

      expect(childEl).not.toBeVisible();

      act(() => {
        fireEvent.click(aggregatorEl);
      });

      expect(childEl).toBeVisible();

      act(() => {
        fireEvent.click(aggregatorEl);
      });

      expect(onClick).toHaveBeenCalledTimes(2);
      expect(childEl).not.toBeVisible();
    });

    it('should not render children when disabled', () => {
      const { queryByTestId } = render(
        <Aggregator {...props} disabled={true}>
          <div data-selector="child">child</div>
        </Aggregator>,
      );
      const childEl = queryByTestId('child');

      expect(childEl).toBeNull();
    });
  });

  describe('accessibility', () => {
    it('should meet accessibility guidelines', async () => {
      const wrapper = renderToHtml(<Aggregator />);
      const actual = await axe(wrapper);
      expect(actual).toHaveNoViolations();
    });
  });
});
