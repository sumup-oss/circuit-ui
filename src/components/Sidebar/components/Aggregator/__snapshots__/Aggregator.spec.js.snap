// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Aggregator styles should render and match snapshot when open 1`] = `
HTMLCollection [
  .circuit-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  height: auto;
  margin: 16px;
  padding: 4px;
  cursor: pointer;
  color: #9DA7B1;
}

.circuit-2:hover {
  color: #D8DDE1;
}

.circuit-0 {
  margin-left: 12px;
}

<div
    class="circuit-2 circuit-3"
    data-testid="aggregator"
  >
    default-icon
    <div
      class="circuit-0 circuit-1"
    >
      Aggregator
    </div>
  </div>,
  .circuit-0 {
  margin: -8px 0 8px 32px;
  list-style: none;
  height: 0;
  position: absolute;
  opacity: 0;
  visibility: hidden;
  overflow: hidden;
  height: calc(32px * );
  position: relative;
  opacity: 1;
  visibility: visible;
  -webkit-transition: height 200ms ease-in-out,opacity 200ms ease-in-out 100ms,visibility 200ms ease-in-out 100ms;
  transition: height 200ms ease-in-out,opacity 200ms ease-in-out 100ms,visibility 200ms ease-in-out 100ms;
}

.circuit-0::after {
  content: '';
  width: 2px;
  background: #3388FF;
  height: 32px;
  border-radius: 1px;
  position: absolute;
  top: calc(32px * 0);
  -webkit-transition: top 200ms ease-in-out;
  transition: top 200ms ease-in-out;
}

.circuit-0::before {
  content: '';
  width: 2px;
  background: #5C656F;
  height: calc(32px * );
  position: absolute;
  top: 0;
  border-radius: 1px;
}

<ul
    class="circuit-0 circuit-1"
  >
    <div
      data-testid="child"
    >
      child
    </div>
  </ul>,
]
`;

exports[`Aggregator styles should render with default styles 1`] = `
HTMLCollection [
  .circuit-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  height: auto;
  margin: 16px;
  padding: 4px;
  cursor: pointer;
  color: #9DA7B1;
}

.circuit-2:hover {
  color: #D8DDE1;
}

.circuit-0 {
  margin-left: 12px;
}

<div
    class="circuit-2 circuit-3"
    data-testid="aggregator"
  >
    default-icon
    <div
      class="circuit-0 circuit-1"
    >
      Aggregator
    </div>
  </div>,
  .circuit-0 {
  margin: -8px 0 8px 32px;
  list-style: none;
  height: 0;
  position: absolute;
  opacity: 0;
  visibility: hidden;
  overflow: hidden;
}

.circuit-0::after {
  content: '';
  width: 2px;
  background: #3388FF;
  height: 32px;
  border-radius: 1px;
  position: absolute;
  top: calc(32px * 0);
  -webkit-transition: top 200ms ease-in-out;
  transition: top 200ms ease-in-out;
}

.circuit-0::before {
  content: '';
  width: 2px;
  background: #5C656F;
  height: calc(32px * );
  position: absolute;
  top: 0;
  border-radius: 1px;
}

<ul
    class="circuit-0 circuit-1"
  >
    <div
      data-testid="child"
    >
      child
    </div>
  </ul>,
]
`;

exports[`Aggregator styles should render with disabled state styles and match the snapshot 1`] = `
.circuit-0 {
  margin-left: 12px;
}

.circuit-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  height: auto;
  margin: 16px;
  padding: 4px;
  cursor: pointer;
  color: #9DA7B1;
  cursor: not-allowed;
  color: #5C656F;
}

<div
  class="circuit-2 circuit-3"
  data-testid="aggregator"
  disabled=""
>
  <svg
    fill="none"
    height="24"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M8 10V7c0-1.031.421-2.132 1.172-2.861A4.059 4.059 0 0112 3c1.06 0 2.078.41 2.828 1.139C15.578 4.869 16 5.969 16 7v3M5 11h14v10H5z"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
    />
  </svg>
  <div
    class="circuit-0 circuit-1"
  >
    Aggregator
  </div>
</div>
`;
